packages/arb-ts/src/lib/utils/arbProvider.ts

```diff
@@ -0,0 +1,246 @@
+
+    const address = this.address.bind(this)
+    const bigNumber = this.bigNumber.bind(this)
+    const data = this.data.bind(this)
+    const hash = this.hash.bind(this)
+    const number = this.number.bind(this)
+    const feeStats = this.feeStats.bind(this)
+    const feeStat = this.feeStat.bind(this)
+    const batchInfo = this.batchInfo.bind(this)
+    const returnCode = this.returnCode.bind(this)
+
+    const arbTransactionFormat = {
+      ...superFormats.transaction,
+
+      l1SequenceNumber: bigNumber,
+      parentRequestId: hash,
+      indexInParent: number,
+      arbType: number,
+      arbSubType: number,
+      l1BlockNumber: number,
+    }
+
+    const arbReceiptFormat = {
+      ...superFormats.receipt,
+      returnData: data,
+      returnCode: returnCode,
+      feeStats: feeStats,
+      batchInfo: Formatter.allowNull(batchInfo, null),
+      l1BlockNumber: number,
+    }
+
+    const feeStatsFormat = {
+      prices: feeStat,
+      unitsUsed: feeStat,
+      paid: feeStat,
+    }
+
+    const feeStatFormat = {
+      l1Transaction: bigNumber,
+      l1Calldata: bigNumber,
+      l2Storage: bigNumber,
+      l2Computation: bigNumber,
+    }
+
+    const batchInfoFormat = {
+      confirmations: number,
+      blockNumber: number,
+      logAddress: address,
+      logTopics: Formatter.arrayOf(hash),
+      logData: data,
+    }
+
+    return {
+      ...superFormats,
+      transaction: arbTransactionFormat,
+      receipt: arbReceiptFormat,
+      feeStats: feeStatsFormat,
+      feeStat: feeStatFormat,
+      batchInfo: batchInfoFormat,
+    }
+  }
+
+  public returnCode(value: any): ReturnCode {
+    const bn = BigNumber.from(value)
+    const returnNum = bn.toNumber()
+    if (!Object.values(ReturnCode).includes(returnNum)) {
+      return logger.throwArgumentError('invalid return code', 'value', value)
+    }
+    return returnNum
+  }
+
+  public feeStat(feeStat: any): FeeStat {
+    return Formatter.check(this.formats.feeStat, feeStat)
+  }
+
+  public feeStats(feeStats: any): FeeStats {
+    return Formatter.check(this.formats.feeStats, feeStats)
+  }
+
+  public batchInfo(batchInfo: any): BatchInfo {
+    return Formatter.check(this.formats.batchInfo, batchInfo)
+  }
+
+  public transactionResponse(transaction: any): ArbTransactionResponse {
+    return super.transactionResponse(transaction) as ArbTransactionResponse
+  }
+
+  public receipt(value: any): ArbTransactionReceipt {
+    return super.receipt(value) as ArbTransactionReceipt
+  }
+}
+
+const getBatch = async (
+  seqNum: BigNumber,
+  l1BlockNumber: number,
+  l1Provider: providers.Provider
+): Promise<Omit<BatchInfo, 'confirmations'> | null> => {
+  const ethNetwork = await getL1Network(l1Provider)
+  const arbNetwork = await getL2Network(ethNetwork.partnerChainIDs[0])
+  const inbox = SequencerInbox__factory.connect(
+    arbNetwork.ethBridge.sequencerInbox,
+    l1Provider
+  )
+
+  const delta = 1000
```

Comment:
I don't we need any delta here; i.e., the actual current l1 Block Number will always be greater than (or maybe equal to) the l1BlockNumber that the Sequencer assigns (since the Sequencer has a delayed view of the L1); the batch at which the transaction ultimately gets submitted will happen at an even greater l1 block number still; thus, I don't see any need to buffer negative/past direction.

Reply:
removed